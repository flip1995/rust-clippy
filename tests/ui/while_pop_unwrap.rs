#![allow(unused)]
#![warn(clippy::while_pop_unwrap)]

struct VecInStruct {
    numbers: Vec<i32>,
    unrelated: String,
}

fn accept_i32(_: i32) {}

fn main() {
    let mut numbers = vec![1, 2, 3, 4, 5];
    while !numbers.is_empty() {
        let number = numbers.pop().unwrap();
    }

    let mut val = VecInStruct {
        numbers: vec![1, 2, 3, 4, 5],
        unrelated: String::new(),
    };
    while !val.numbers.is_empty() {
        let number = val.numbers.pop().unwrap();
    }

    while !numbers.is_empty() {
        accept_i32(numbers.pop().unwrap());
    }

    while !numbers.is_empty() {
        accept_i32(numbers.pop().expect(""));
    }

    // This should not warn. It "conditionally" pops elements.
    while !numbers.is_empty() {
        if true {
            accept_i32(numbers.pop().unwrap());
        }
    }

    // This should also not warn. It conditionally pops elements.
    while !numbers.is_empty() {
        if false {
            continue;
        }
        accept_i32(numbers.pop().unwrap());
    }
}
