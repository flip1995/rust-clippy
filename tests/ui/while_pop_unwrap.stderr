error: you seem to be trying to pop elements from a `Vec` in a loop
  --> $DIR/while_pop_unwrap.rs:14:22
   |
LL |     while !numbers.is_empty() {
   |            ------------------ help: try: `while let Some(number) = numbers.pop()`
LL |         let number = numbers.pop().unwrap();
   |                      ^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this while loop can be written in a more idiomatic way
   = note: `-D clippy::while-pop-unwrap` implied by `-D warnings`

error: you seem to be trying to pop elements from a `Vec` in a loop
  --> $DIR/while_pop_unwrap.rs:22:22
   |
LL |     while !val.numbers.is_empty() {
   |            ---------------------- help: try: `while let Some(number) = val.numbers.pop()`
LL |         let number = val.numbers.pop().unwrap();
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this while loop can be written in a more idiomatic way

error: you seem to be trying to pop elements from a `Vec` in a loop
  --> $DIR/while_pop_unwrap.rs:26:20
   |
LL |     while !numbers.is_empty() {
   |            ------------------ help: try: `while let Some(element) = numbers.pop()`
LL |         accept_i32(numbers.pop().unwrap());
   |                    ^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this while loop can be written in a more idiomatic way

error: you seem to be trying to pop elements from a `Vec` in a loop
  --> $DIR/while_pop_unwrap.rs:30:20
   |
LL |     while !numbers.is_empty() {
   |            ------------------ help: try: `while let Some(element) = numbers.pop()`
LL |         accept_i32(numbers.pop().expect(""));
   |                    ^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this while loop can be written in a more idiomatic way

error: aborting due to 4 previous errors

